File: leetcode/medium/64-min-path/iter-dp.py
"""
  @author Vortexx2
  Problem 64 - Minimum Path Sum
  @approach Using a row by row iterative DP approach

  Runtime - 171 ms
  Memory Usage - 14.7 MB
"""
from typing import List


class Solution:
  def minPathSum(self, grid: List[List[int]]) -> int:

    rows, cols = len(grid), len(grid[0])

    def firstRow() -> List[int]:
      row = [0] * cols
      row[0] = grid[0][0]

      for i in range(1, cols):
        row[i] = row[i - 1] + grid[0][i]

      return row

    row = firstRow()

    for i in range(1, rows):
      for j in range(0, cols):

        if j:
          row[j] = min(row[j - 1], row[j])

        row[j] += grid[i][j]

    return row[cols - 1]

File: leetcode/medium/64-min-path/iter-dp.cpp
/**
 * @file iter-dp.cpp
 * @author Vortexx2
 * @brief Problem 64 - Minimum Path Sum
 * @date 03-08-2022
 * 
 * Runtime - 16 ms O(m * n)
 * Memory Usage - 9.7 MB O(n) (n = number of columns)
 */
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int minPathSum(vector<vector<int>>& grid) {
    int rows = grid.size(), cols = grid[0].size();

    auto row = firstRow(grid);

    for (int i = 1; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (j) row[j] = min(row[j], row[j - 1]);

        row[j] += grid[i][j];
      }
    }

    return row[cols - 1];
  }

  vector<int> firstRow(vector<vector<int>>& grid) {
    vector<int> row(grid[0].size());
    row[0] = grid[0][0];
    for (int i = 1; i < grid[0].size(); i++) {
      row[i] = row[i - 1] + grid[0][i];
    }

    return row;
  }
};